packages/std/src/math/signed_decimal.rs

```diff
@@ -0,0 +1,3060 @@
+                Self(
+                    atomics
+                        .checked_mul(factor)
+                        .map_err(|_| SignedDecimalRangeExceeded)?,
+                )
+            }
+            Ordering::Equal => Self(atomics),
+            Ordering::Greater => {
+                let digits = decimal_places - (Self::DECIMAL_PLACES); // No overflow because decimal_places > DECIMAL_PLACES
+                if let Ok(factor) = TEN.checked_pow(digits) {
+                    Self(atomics.checked_div(factor).unwrap()) // Safe because factor cannot be zero
+                } else {
+                    // In this case `factor` exceeds the Int128 range.
+                    // Any Int128 `x` divided by `factor` with `factor > Int128::MAX` is 0.
+                    // Try e.g. Python3: `(2**128-1) // 2**128`
+                    Self(Int128::zero())
+                }
+            }
+        })
+    }
+
+    /// Returns the ratio (numerator / denominator) as a SignedDecimal
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use cosmwasm_std::SignedDecimal;
+    /// assert_eq!(
+    ///     SignedDecimal::from_ratio(1, 3).to_string(),
+    ///     "0.333333333333333333"
+    /// );
+    /// ```
+    pub fn from_ratio(numerator: impl Into<Int128>, denominator: impl Into<Int128>) -> Self {
+        match SignedDecimal::checked_from_ratio(numerator, denominator) {
+            Ok(value) => value,
+            Err(CheckedFromRatioError::DivideByZero) => {
+                panic!("Denominator must not be zero")
+            }
+            Err(CheckedFromRatioError::Overflow) => panic!("Multiplication overflow"),
+        }
+    }
+
+    /// Returns the ratio (numerator / denominator) as a SignedDecimal
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use cosmwasm_std::{SignedDecimal, CheckedFromRatioError};
+    /// assert_eq!(
+    ///     SignedDecimal::checked_from_ratio(1, 3).unwrap().to_string(),
+    ///     "0.333333333333333333"
+    /// );
+    /// assert_eq!(
+    ///     SignedDecimal::checked_from_ratio(1, 0),
+    ///     Err(CheckedFromRatioError::DivideByZero)
+    /// );
+    /// ```
+    pub fn checked_from_ratio(
+        numerator: impl Into<Int128>,
+        denominator: impl Into<Int128>,
+    ) -> Result<Self, CheckedFromRatioError> {
+        let numerator: Int128 = numerator.into();
+        let denominator: Int128 = denominator.into();
+        match numerator.checked_multiply_ratio(Self::DECIMAL_FRACTIONAL, denominator) {
+            Ok(ratio) => {
+                // numerator * DECIMAL_FRACTIONAL / denominator
+                Ok(SignedDecimal(ratio))
+            }
+            Err(CheckedMultiplyRatioError::Overflow) => Err(CheckedFromRatioError::Overflow),
+            Err(CheckedMultiplyRatioError::DivideByZero) => {
+                Err(CheckedFromRatioError::DivideByZero)
+            }
+        }
+    }
+
+    /// Returns `true` if the number is 0
+    #[must_use]
```

Comment:
Is `#[must_use]` neccessary?

Reply:
They are never strictly necessary, but it doesn't really hurt to have them and the rust standard library also has them on functions that return bools. See for example here: https://doc.rust-lang.org/src/core/num/mod.rs.html#1052